File: algorithms/code/leetcode/lc101_symmetric_tree/LC101SymmetricTree.java
package leetcode.lc101_symmetric_tree;

import java.util.ArrayList;

/**
 * https://leetcode.com/problems/symmetric-tree/ #easy
 */
public final class LC101SymmetricTree {
    public static class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;

        TreeNode(final int x) {
            val = x;
        }
    }

    private static class TreeNodePair {
        public TreeNode node1;
        public TreeNode node2;

        TreeNodePair(final TreeNode node1Value, final TreeNode node2Value) {
            node1 = node1Value;
            node2 = node2Value;
        }
    }

    private enum NullCheckResult {
        ALL_NULL, SOME_NULL, NONE_NULL
    }

    private static NullCheckResult nullCheck(final TreeNode tree1, final TreeNode tree2) {
        boolean allNull = tree1 == null && tree2 == null;
        boolean noneNull = tree1 != null && tree2 != null;
        boolean someNull = !allNull && !noneNull;
        if (allNull) {
            return NullCheckResult.ALL_NULL;
        }
        if (someNull) {
            return NullCheckResult.SOME_NULL;
        }
        return NullCheckResult.NONE_NULL;
    }

    public boolean isSymmetric(final TreeNode root) {
        if (root == null) {
            return true;
        }
        ArrayList<TreeNodePair> queue = new ArrayList<>();
        queue.add(new TreeNodePair(root, root));
        while (queue.size() > 0) {
            TreeNodePair pair = queue.remove(0);
            TreeNode node1 = pair.node1;
            TreeNode node2 = pair.node2;
            if (node1.val != node2.val) {
                return false;
            }
            NullCheckResult checkLeftToRight = nullCheck(node1.left, node2.right);
            if (checkLeftToRight == NullCheckResult.NONE_NULL) {
                queue.add(new TreeNodePair(node1.left, node2.right));
            } else {
                if (checkLeftToRight == NullCheckResult.SOME_NULL) {
                    return false;
                }
            }
            NullCheckResult checkRightToLeft = nullCheck(node1.right, node2.left);
            if (checkRightToLeft == NullCheckResult.NONE_NULL) {
                queue.add(new TreeNodePair(node1.right, node2.left));
            } else {
                if (checkRightToLeft == NullCheckResult.SOME_NULL) {
                    return false;
                }
            }
        }
        return true;
    }
}
package leetcode.lc101_symmetric_tree;

import org.junit.jupiter.api.Test;

import leetcode.lc101_symmetric_tree.LC101SymmetricTree.TreeNode;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public final class LC101SymmetricTreeTests {
    @Test
    public void testSymmetric() throws Exception {
        TreeNode n0 = new TreeNode(0);
        TreeNode n1a = new TreeNode(1);
        TreeNode n1b = new TreeNode(1);
        n0.left = n1a;
        n0.right = n1b;
        n1a.left = null;
        n1a.right = null;
        n1b.left = null;
        n1b.right = null;
        assertTrue(new LC101SymmetricTree().isSymmetric(n0));
    }

    @Test
    public void testAsymmetric() throws Exception {
        TreeNode n0 = new TreeNode(0);
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        n0.left = n1;
        n0.right = n2;
        n1.left = null;
        n1.right = null;
        n2.left = null;
        n2.right = null;
        assertFalse(new LC101SymmetricTree().isSymmetric(n0));
    }

    @Test
    public void testEmpty() throws Exception {
        assertTrue(new LC101SymmetricTree().isSymmetric(null));
    }

    @Test
    public void testLeft() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        an0.left = an1;
        assertFalse(new LC101SymmetricTree().isSymmetric(an0));
    }

    @Test
    public void testRight() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        an0.right = an1;
        assertFalse(new LC101SymmetricTree().isSymmetric(an0));
    }

    @Test
    public void testOther() throws Exception {
        TreeNode n2 = new TreeNode(2);
        TreeNode n3l = new TreeNode(3);
        TreeNode n3r = new TreeNode(3);
        TreeNode n4ll = new TreeNode(4);
        TreeNode n5 = new TreeNode(5);
        TreeNode n4rr = new TreeNode(4);
        n2.left = n3l;
        n2.right = n3r;
        n3l.left = n4ll;
        n3l.right = n5;
        n3r.right = n4rr;
        assertFalse(new LC101SymmetricTree().isSymmetric(n2));
    }
}

File: algorithms/code/leetcode/lc101_symmetric_tree/lc101_symmetric_tree.py
#!/usr/bin/env python3
# https://leetcode.com/problems/symmetric-tree/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: 'TreeNode | None' = None,
        right: 'TreeNode | None' = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode | None) -> bool:
        if root is None:
            return True
        queue: list[tuple[TreeNode, TreeNode]] = []
        queue.append((root, root))
        while queue:
            node1, node2 = queue.pop(0)
            if node1.val != node2.val:
                return False
            node1_left, node2_right = node1.left, node2.right
            if node1_left is not None and node2_right is not None:
                queue.append((node1_left, node2_right))
            elif not (node1_left is None and node2_right is None):
                return False
            node1_right, node2_left = node1.right, node2.left
            if node1_right is not None and node2_left is not None:
                queue.append((node1_right, node2_left))
            elif not (node1_right is None and node2_left is None):
                return False
        return True


class TestCode(unittest.TestCase):
    def test_symmetric(self) -> None:
        n0 = TreeNode(0)
        n1a = TreeNode(1)
        n1b = TreeNode(1)
        n0.left = n1a
        n0.right = n1b
        n1a.left = None
        n1a.right = None
        n1b.left = None
        n1b.right = None
        self.assertTrue(Solution().isSymmetric(n0))

    def test_asymmetric(self) -> None:
        n0 = TreeNode(0)
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n0.left = n1
        n0.right = n2
        n1.left = None
        n1.right = None
        n2.left = None
        n2.right = None
        self.assertFalse(Solution().isSymmetric(n0))

    def test_empty(self) -> None:
        self.assertTrue(Solution().isSymmetric(None))

    def test_left(self) -> None:
        an0 = TreeNode(0)
        an1 = TreeNode(1)
        an0.left = an1
        self.assertFalse(Solution().isSymmetric(an0))

    def test_right(self) -> None:
        an0 = TreeNode(0)
        an1 = TreeNode(1)
        an0.right = an1
        self.assertFalse(Solution().isSymmetric(an0))

    def test_other(self) -> None:
        n2 = TreeNode(2)
        n3l = TreeNode(3)
        n3r = TreeNode(3)
        n4ll = TreeNode(4)
        n5 = TreeNode(5)
        n4rr = TreeNode(4)
        n2.left = n3l
        n2.right = n3r
        n3l.left = n4ll
        n3l.right = n5
        n3r.right = n4rr
        self.assertFalse(Solution().isSymmetric(n2))
